plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx

```diff
@@ -19,9 +20,35 @@ import InspectorControls from './inspector-controls';
 import InspectorAdvancedControls from './inspector-advanced-controls';
 import ToolbarControls from './toolbar-controls';
 
-const ProductCollectionContent = (
-	props: ProductCollectionEditComponentProps
-) => {
+const usePreviewState = ( handlePreviewState ) => {
+	// I have implemented this internal state to handle the preview state.
+	// As you can see it contains isPreview and previewMessage.
+	// - isPreview is a boolean to check if the block is in preview mode.
+	// - previewMessage is a string to display the preview message in tooltip.
+	const [ previewState, setPreviewState ] = useState( {
+		isPreview: false,
+		previewMessage: '',
+	} );
+
+	// Running handlePreviewState function provided by Collection, if it exists.
+	useEffect( () => {
+		handlePreviewState?.( {
+			setPreviewState,
+		} );
+
+		// We want this to run only once, adding deps will cause performance issues.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [] );
```

Comment:
Lets say I wrote some async handler, that starts with preview on and then switches it off, if I do that there will be a flash with this approach, we render the component with **no preview** and then handler will **turn preview on (after mount)**, and then lets say 500ms later it will **turn off again**.

I think the API needs to support setting the initial state to ensure this flash wouldn't happen. some would want to start with it on, and others would want to start with it off. I'm not sure how to make the api for that very nice and I don't have an immediate idea of how to support that nicely with useState, but I think the concern is relevant.

Reply:
@kmanijak that won't work, there will still be a flash because we don't call setPreviewState for the first time until the component first mounts..

Maybe its fast enough to be rendered not noticeable here though? but there will be an initial redundant re-render.

Reply:
> Maybe its fast enough to be rendered not noticeable here though? but there will be an initial redundant re-render.

Yes, it should be fast enough but I agree - it's not optimal. On the other hand, it's a tradeoff between optimization and potentially more complex API. Let's see if we have some idea of how to expose a way to set the initial value! Maybe it won't be too complex and I'm simply wrong about that.

Reply:
I think on top of what @kmanijak shared in https://github.com/woocommerce/woocommerce/pull/45703#discussion_r1537713783 comment, we can use `useLayoutEffect` instead of `useEffect`. This way we can avoid the initial flash. 

**`useEffect` vs. `useLayoutEffect`**

- `useEffect`: Runs asynchronously after the DOM has been updated. If you're making changes to the DOM or triggering a re-render from `useEffect`, those changes will not be applied immediately but in the next tick, potentially leading to visible flashes or janky updates if there's a significant difference between the initial and subsequent states.
- `useLayoutEffect`: Runs synchronously after DOM mutations but before the browser has painted the changes to the screen. This timing allows you to make adjustments to the DOM or execute logic based on the DOM state without the user seeing any intermediate states, effectively preventing visual flashes or flickers caused by updates.

@samueljseay What do you think? Am I missing something? 🙂

Reply:
@imanish003 great idea, I hadn't considered using that. Yes I think since we this is not resource intensive and won't be called often that we can use `useLayoutEffect`. Good thinking!

Reply:
Awesome, I will make the changes 🙌🏻